File: waboranggae/android-native/app/src/main/java/kr/co/waboranggae/nativepilot/data/RouteInsertion.kt
package kr.co.waboranggae.nativepilot.data
import kotlin.math.*
data class InsertionCandidate(val place:Place,val index:Int,val detourKm:Double)
fun routeDistance(a:Coordinate,b:Coordinate):Double{
 val rad=Math.PI/180;val dy=(b.latitude-a.latitude)*rad;val dx=(b.longitude-a.longitude)*rad
 val v=sin(dy/2).pow(2)+cos(a.latitude*rad)*cos(b.latitude*rad)*sin(dx/2).pow(2)
 return 6371*2*atan2(sqrt(v),sqrt(max(0.0,1-v)))
}
fun nearbyAdditions(route:List<Place>,pool:List<Place>,origin:Coordinate?,excluded:Set<String> = emptySet()):List<InsertionCandidate>{
 val used=route.map{it.id}.toSet()+excluded
 return pool.distinctBy{it.id}.filter{it.id !in used && it.category!="station" && route.none{p->p.name.replace(" ","")==it.name.replace(" ","")}}.mapNotNull{place->
  val point=place.coordinate()?:return@mapNotNull null
  (0..route.size).mapNotNull{index->
   val before=if(index==0)origin else route[index-1].coordinate()
   val after=route.getOrNull(index)?.coordinate()
   if(before==null&&after==null)return@mapNotNull null
   val extra=(before?.let{routeDistance(it,point)}?:0.0)+(after?.let{routeDistance(point,it)}?:0.0)-(if(before!=null&&after!=null)routeDistance(before,after)else 0.0)
   InsertionCandidate(place,index,max(0.0,extra))
  }.minByOrNull{it.detourKm}
 }.sortedBy{it.detourKm}
}

File: waboranggae/src/domain/routeInsertion.ts
type Point={latitude?:number;longitude?:number};
export function routeDistance(a:Point,b:Point){
 if(!Number.isFinite(a.latitude)||!Number.isFinite(a.longitude)||!Number.isFinite(b.latitude)||!Number.isFinite(b.longitude))return Infinity;
 const rad=Math.PI/180,dy=(b.latitude!-a.latitude!)*rad,dx=(b.longitude!-a.longitude!)*rad;
 const v=Math.sin(dy/2)**2+Math.cos(a.latitude!*rad)*Math.cos(b.latitude!*rad)*Math.sin(dx/2)**2;
 return 6371*2*Math.atan2(Math.sqrt(v),Math.sqrt(Math.max(0,1-v)));
}
/** Public place coordinates only. Straight-line detour ranking, not verified travel time. */
export function bestInsertion(route:Point[],candidate:Point,origin?:Point){
 let index=route.length,detourKm=Infinity;
 for(let i=0;i<=route.length;i++){
  const before=i?route[i-1]:origin,after=route[i];
  const extra=(before?routeDistance(before,candidate):0)+(after?routeDistance(candidate,after):0)-(before&&after?routeDistance(before,after):0);
  if(Number.isFinite(extra)&&extra<detourKm){index=i;detourKm=Math.max(0,extra);}
 }
 return {index,detourKm};
}
export function nearbyAdditions<T extends Point&{id:string;name:string}>(route:T[],pool:T[],origin?:Point,excluded:string[]=[]){
 const used=new Set([...route.map(p=>p.id),...excluded]),names=new Set(route.map(p=>p.name.replace(/\s/g,'')));
 return [...new Map(pool.map(p=>[p.id,p])).values()].filter(p=>!used.has(p.id)&&!names.has(p.name.replace(/\s/g,'')))
  .map(place=>({place,...bestInsertion(route,place,origin)})).filter(p=>Number.isFinite(p.detourKm)).sort((a,b)=>a.detourKm-b.detourKm||a.place.name.localeCompare(b.place.name,'ko'));
}
